### PR TITLE
Extension refactoring

### LINKED ISSUE #732: Migrator uses non studly cased namespace for extensions
The [Migrator](https://github.com/flarum/core/blob/master/src/Database/Migrator.php#L249) class is formatting the namespace of extensions as lowercase:

```
exception 'ReflectionException' with message 'Class flagrow\image\upload\Migration\CreateFlagrowImagesTable does not exist'
```
When the class is:

```php
<?php namespace Flagrow\ImageUpload\Migration;

/*
* This file is part of image-upload.
*
* (c) Flagrow
*
* For the full copyright and license information, please view the license.md
* file that was distributed with this source code.
*/

use Flarum\Database\AbstractMigration;
use Illuminate\Database\Schema\Blueprint;

class CreateFlagrowImagesTable extends AbstractMigration
{
```

### PR DESCRIPTION
**Pull request by Luceos**
Fixes #732 

This PR fixes the above issue as well as prepare for more things to happen with Extensions as a whole.

This pr does some stuff:

- ExtensionManager::getInfo is obsoleted.
- ExtensionManager::getExtensions and related methods added.
- Created an Extension class that does most of the "tough" work now.
- Rewrote any method that needs a whole Extension object to take it as argument.
- Related Extension events now take the whole object as argument.
- Refactor the admin ExtensionsPage to take the getExtensions results better.. Or create a filter before sending it to the view.

### CODE REVIEW

**Comment by tobscure** [comment]
Looking good! Is this based on the latest master? There are a lot of StyleCI problems...

**Comment by Luceos** [comment]
I thought so @tobscure . Did you merge your style-ci changes already in latest master? I've pulled in the latest changes regularly, but not yet those by @franzliedke from yesterday evenening. We were both working at the same time, that's why travis is giving issues.

**Comment by tobscure** [comment]
My bad, I didn't merge them in yet. Carry on.

**Comment by Luceos** [comment]
@flarum/core please review. Need js admin recompilation before testing.

Frontend now supports new list too. Might add an Event at some points, but let's do that later.

**Comment by Luceos** [comment]
travis fixed ;)

**Comment by Luceos** [comment]
@flarum/core ;)

File: js/admin/src/components/ExtensionsPage.js [hunk 1]
```diff
@@ -11,7 +11,7 @@ import listItems from 'flarum/helpers/listItems';
 
 export default class ExtensionsPage extends Component {
   view() {
-    const extensions = Object.keys(app.extensions).map(id => app.extensions[id]);
+    const extensions = app.extensions;
```
Comment (franzliedke) [inline js/admin/src/components/ExtensionsPage.js]: I'd say just use `app.extensions` throughout the method.

File: src/Extension/Extension.php [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+            if ($file = Arr::get($icon, 'image')) {
+                $file = $this->path . '/' . $file;
+
+                if (file_exists($file)) {
+                    $mimetype = pathinfo($file, PATHINFO_EXTENSION) === 'svg'
+                        ? 'image/svg+xml'
+                        : finfo_file(finfo_open(FILEINFO_MIME_TYPE), $file);
+                    $data     = file_get_contents($file);
+
+                    $icon['backgroundImage'] = 'url(\'data:' . $mimetype . ';base64,' . base64_encode($data) . '\')';
```
Comment (franzliedke) [inline src/Extension/Extension.php]: Can we just expose the data here and do this CSS-specific formatting where it's used?

Comment (Luceos) [inline src/Extension/Extension.php]: I'm not sure. Like a custom component? Does make sense.. @tobscure ?

Comment (Luceos) [inline src/Extension/Extension.php]: @franzliedke maybe push this to a seperate item? This pr is already monstruous and keeps all previous functionality _intact_.

Comment (tobscure) [inline src/Extension/Extension.php]: Agreed, let's talk about this in another issue

**Comment by tobscure** [comment]
Haven't had time to review yet (will hopefully do on the weekend) but a heads up that you will need to squash the commits :) Happy to try and help via Gitter if you need.

**Comment by Luceos** [comment]
@tobscure thanks for offering the assistance, I'm so happy that I even solved the hardest part (frontend) ;) As said there are a few minor details that need fixing, like squashing and @franzliedke 's comment on the direct accessor of extension. I'll finish it up today.

**Comment by Luceos** [comment]
Really need to get used to squashing :laughing: Squashed it to two for now, @tobscure

**Comment by franzliedke** [comment]
We're pretty happy with this. :)

So what's the status? This is still marked as WIP. Toby is planning to merge the StyleCI changes soon, so once that's done, you can rebase on master and we'll merge. :)

**Comment by Luceos** [comment]
@flarum/core you can merge now, it worked perfectly on my env. You can then merge the latest PR from styleci. It doesn't really matter. It's not a wip anymore ;)

**Comment by franzliedke** [comment]
Well, you still have two commits, and the change to .travis.yml is already in master.

It looks good, though. I'll let @tobscure have a final look and do the merge - he can always squash-merge, too. :)

Thanks again, @Luceos !

**Comment by tobscure** [comment]
@franzliedke How do you squash merge? Wouldn't it make me the author of the commit rather Luceos?

I think this is a good opportunity for @Luceos to learn how to squash once and for all :P again, I'm happy to offer assistance if needed!

**Comment by franzliedke** [comment]
You can do `git merge --squash his-branch`. And then you can even change the author, by amending the commit with `git commit --amend --author="some body <his@email.com>"`.

See https://gist.github.com/sindresorhus/2630026.

**Comment by tobscure** [comment]
Very good to know!

**Comment by Luceos** [comment]
I'll try to squash again.

**Comment by Luceos** [comment]
Would love to have one contribution actually being made on my name ;) All previous ones have been added manually so far :(

**Comment by franzliedke** [comment]
No worries. Whatever happens, we'll be able to commit with your name, if everything else fails. ;)

**Comment by Luceos** [comment]
I feel like a donkey, everytime I squash the latest 2, it says it's not up to date with the remote. Pulling results in a merge which also shows up as commit and I am not able to squash those :disappointed:

**Comment by franzliedke** [comment]
It's okay, we can take care of it when merging, shall we?

**Comment by Luceos** [comment]
Sure, let's get this thing of my todo once and for all ;)

**Comment by franzliedke** [comment]
I squash-merged this in https://github.com/flarum/core/commit/191589e2b1a633d92ccac57e49b8805dd23252cd.

Thanks a lot!

**Pull request closed without merging**